File: src/levels/L24-un-solo-pago/ts/strategy.ts
// Strategy "por si acaso": una interfaz, un registro… y una sola implementación.
const IVA = 0.16

interface PaymentMethod {
  charge(total: number): void
}

class Cashier {
  constructor(
    private barista: Barista,
    private methods: Record<string, PaymentMethod>,
  ) {}

  // region: Cashier#charge
  charge(order: Order) {
    this.methods['card'].charge(order.price * (1 + IVA))
    this.barista.prepare(order)
  }
  // endregion
}

// region: CardPayment#charge
class CardPayment implements PaymentMethod {
  charge(total: number) { CardTerminal.charge(total) }
}
// endregion

// region: wiring
const cashier = new Cashier(new Barista(new Counter()), { card: new CardPayment() })
new OrderTaker(cashier).take('latte', 50)
// endregion

File: src/levels/L24-un-solo-pago/strategy.rb
# Strategy "por si acaso": una interfaz, un registro… y una sola implementación.
IVA = 0.16

class Cashier
  def initialize(barista, methods)
    @barista = barista
    @methods = methods
  end

  # region: Cashier#charge
  def charge(order)
    @methods.fetch(:card).charge(order.price * (1 + IVA))
    @barista.prepare(order)
  end
  # endregion
end

# region: CardPayment#charge
class CardPayment
  def charge(total) = CardTerminal.charge(total)
end
# endregion

# region: wiring
cashier = Cashier.new(Barista.new(Counter.new), { card: CardPayment.new })
OrderTaker.new(cashier).take(:latte, 50)
# endregion
